dev-build/app/AppLoader.scala

```diff
@@ -52,6 +56,11 @@ trait Controllers
     with FrontendComponents
     with CricketControllers {
   self: BuiltInComponents =>
+
+  // Required for EmailSignupController
+  lazy val newsletterApi = wire[NewsletterApi]
+  lazy val emailEmbedAgent = wire[EmailEmbedAgent]
```

Comment:
Let's talk about this one, I think there might be a way to avoid this and the override below but I can't quite pin point it just yet.

Reply:
Okay I've done more thinking and jogged my memory. I think the most reasonable middle ground is this.

```suggestion
  def emailEmbedAgent: EmailEmbedAgent
```

This essentially says that in order to create a new instance of the trait `Controllers`, the instance will need to have defined a method or attribute named `emailEmbedAgent` returning the type `EmailEmbedAgent` (And it will be enforced by the compiler). That's enough for the wiring macro to pass `emailEmbedAgent` to the constructor of `EmailSignupController`

And that works because you provide such an attribute in the trait `AppComponents`, so at the time the object is instantiated (line 42, in the `buildComponents` method), the compiler will essentially mash all the trait into a single namespace and check that for each `def` that isn't implemented, there is one and only one matching implementation (the `val`s).

To give a bit more background: aeons ago - when I worked on the frontend migration to play 2.6 - I thought about grouping controllers by their application [like so](https://github.com/guardian/frontend/pull/13593/files#diff-d5b2a2617fa860ef98c2e8f2a5e3c33f78ef74375a93fe4ae6449819fffe5e77). It seemed handy because we could then use these groups in both the application and the dev-build. What I didn't foresee was the fact Play's new way of doing things would require more than grouping controllers, as the controllers themselves were more and more going to need other components during their construction (services, api-access etc).

Bottom line is: my refactor didn't age very well 😅